### PR TITLE
Ensure .pagination.next is not present for oldest year

### LINKED ISSUE #5: Edge case issue
Initially reported on https://github.com/nodejs/nodejs.org/issues/2422

The current tests cover one similar case, but I think we need a more extended test case. In nodejs.org, we have `blog/index.md`. Maybe this will trigger the issue here too.

PR which works around the problem: https://github.com/nodejs/nodejs.org/pull/2564

### PR DESCRIPTION
**Pull request by XhmikosR**
@phillipj: this seems to trigger the failure which is the issue we are hitting upstream. Not sure if I'm missing something, though.

Fixes #5

### CODE REVIEW

**Comment by phillipj** [comment]
Really nice! This is very valuable when trying to grasp what on earth is causing this bug in the first place 💯

I'll do my best to squeeze in a thorough look tomorrow and see if I remember much of how this works.

**Comment by XhmikosR** [comment]
Friendly ping @phillipj :)

Maybe we should explicitly set this to null or undefined when we are at the last year?

**Comment by phillipj** [comment]
Yo, thanks for the ping!

I checked out this branch a couple of days ago and tried to grasp what this code actually does.. I'll be honest and say the cloning involved in [./index.js#L42](https://github.com/phillipj/metalsmith-yearly-pagination/blob/master/index.js#L42) and below wasn't really obvious at first glance.

Your suggestion of always setting `.next` to `undefined|null` (or not set it at all?) when we're looping through posts for the last year, sounds sensible. Haven't tried it myself yet, but guess that's what would be most sensible, especially when seeing what you discovered in the nodejs.org repo.

**Comment by XhmikosR** [comment]
I'm also getting lost in that clone part, so I couldn't make a lot of sense of it.

https://github.com/phillipj/metalsmith-yearly-pagination/blob/3a305bc1c85ac03a2913f5bbe4fe92bc4c910015/index.js#L48-L55

Basically when we are at `years[years.length - 1]` I think we should not set it or null it.

**Comment by phillipj** [comment]
👍 

If not setting `.next` make the tests added happy, you could probably verify if that makes the nodejs.org issues go away with `npm link`ing locally?

I'm more than happy to push a new version this weekend if needed.

**Comment by XhmikosR** [comment]
I don't have a working patch. I was hoping you could help me out here because I don't quite get all that cloning.

Before releasing anything, we'll verify it of course :)

**Comment by phillipj** [comment]
Alrighty, I'll give it a try locally and see it affects the added tests 🤞

**Comment by phillipj** [comment]
Will try more later, first naive attempt didn't yield expected results I'm afraid.

**Comment by phillipj** [comment]
Found a way to make all the tests pas.. You okey with me pushing a commit to your branch?

**Comment by phillipj** [comment]
Pushed my trivial fix (incl/your tests) to [fix-next-pagination-in-oldest-year](https://github.com/phillipj/metalsmith-yearly-pagination/commits/fix-next-pagination-in-oldest-year) in case you want to do some early stage verification.

**Comment by XhmikosR** [comment]
You could push here ofc :)

Anyway, I tested the branch and seems to be fine upstream. Could we miss something else?

**Comment by XhmikosR** [comment]
Only thing that bugs me is that we shouldn't set the property in the first place. Not sure how much it would complicate things, but it feels the right thing instead of deleting the property :)

**Comment by phillipj** [comment]
Myeah, totally see your point there. Although deleting the `.next` property _works_, I agree it's not ideal.

When trying out other approaches, I ended up feeling like there's some mutable references that really caught me off guard and therefore ended up causing more headache than I cared to understand.

Which obviously raises an obvious need to refactor the inner workings a bit, so that our future selfs understands what the heck is going on.

I'm willing to try a last shot at understand & refactoring the cloning and reference parts, but I got other things on my table today as well, so I won't spend too much time on it. In worst case, we've now got a plausible solution by deleting the `.pagination.next`.

**Comment by XhmikosR** [comment]
Agreed. I'll try experimenting with this and let you know. But perhaps you
should merge this and your patch? Just don't release it yet

**Comment by phillipj** [comment]
Ohohoho I finally get it!

Personally see this as a bug from multiple levels of mutable references, where we set `.pagination.next` in one iteration of `years.forEach()` and we get that value with us in the next iteration as well. My ramblings probably doesn't make sense, but I'll get a fix in place to ensure this mindbending experience doesn't hit us as hard again later.

Also sadly have realised there's no way around having to mutate references being provided by metalsmith, as it seems that's really the "API" metalsmith plugins use; alter the posts & collections in place. My brain would really appreciate if we could avoid mutating references like that, but that seems impossible with how metalsmith works as far as I see things.

File: index.js [hunk 1]
```diff
@@ -38,20 +38,22 @@ function paginate (filePath, collection, fileName, files, iteratee) {
...
+        const currentYear = cloneDeepWith(origFile, (value) => {
             if (Buffer.isBuffer(value)) {
                 return value.slice();
             }
         });
 
-        last.pagination.next = clone;
-        clone.pagination.year = year;
-        clone.pagination.prev = last;
-        clone.pagination.posts = posts.map(iteratee);
```
Comment (phillipj) [inline index.js:50]: Previously, we mutated the `.pagination` object which was a copy of last year's pagination object -- that included the `.pagination.next` property.

Now that we're always re-creating a new `.pagination` object, we're ensuring that the last year's `.pagination.next` isn't present anymore.

That property is only set in the _next iteration_, which doesn't happen for the oldest year -- voila, bug fixed.

**Comment by phillipj** [comment]
Rebased with master and included a better fix than the `delete ..` fix discussed earlier.

Also included a couple of variable renamings which surely helped my brain grok what was going on.

Shout if you don't really agree this was much better than `delete`ing, or else I'll merge within a day or two.

**Comment by XhmikosR** [comment]
LGTM as far as I can tell. I tested it with nodejs.org and works as expected. Good job!

**Comment by XhmikosR** [comment]
BTW this branch seems slightly slower, just FYI.

**Comment by XhmikosR** [comment]
Scratch my comment about this being slower; we run more tests :)

**Comment by phillipj** [comment]
Cool! Thx a lot for your patience and pushing for getting this bug fixed in the right place, rather than in using projects 👍

**Comment by XhmikosR** [comment]
Let me know when you release a new version so that update it upstream :)

**Comment by phillipj** [comment]
v4.0.0 is available on npm now 🚀
